### PR TITLE
Implement basic support for reader conditionals

### PR DESCRIPTION
**Pull request by vemv**
Hey there,

here's my stab at https://github.com/gfredericks/how-to-ns/issues/9 .

Probably interspersing code with `reader-conditional?` isn't very elegant, so I'm open to change it to sth else. Thought I'd be cautious to start small :)

Strictly speaking , reader conditionals can appear anywhere in the code, but this PR only deals with likely choices from the user.

There's no attempt whatsoever at formatting the contents of the reader conditional. Can eventually do.

WDYT?

Victor

### CODE REVIEW

File: how-to-ns/src/com/gfredericks/how_to_ns.clj [hunk 1]
```diff
@@ -99,24 +99,41 @@
...
     (->> all-classes
-         (map split-symbol)
+         (map (fn [x]
+                (if (reader-conditional? x)
+                    [x]
+                    (split-symbol x))))
          (group-by first)
-         (sort)
+         (sort-by (fn [[x]]
+                    (str x)))
```
Comment (gfredericks) [inline how-to-ns/src/com/gfredericks/how_to_ns.clj]: `x` might be a reader conditional here, is that correct?

according to my repl, `str` on a reader conditional gives something like `clojure.lang.ReaderConditional@5044f1cd`, which seems like it would yield a nondeterministic sort here; `pr-str` would likely do better; I'd like something that involves the contents of the conditional, but I don't know the best way to do that; if you know what order the kv-pairs inside it would come in, you could sort based on the first value.

```

**Review by gfredericks — COMMENTED** [review]
I mentioned a sorting issue or two, and I also started thinking about the align-clauses issue, and how weird it will look to have the reader conditionals sprinkled throughout the otherwise-aligned clauses.

Note that you can use the `:form` and `:splicing?` keywords to extract the data parts of a reader conditional, and the `reader-conditional` function to reconstruct one.

This means we could possibly define some kind of canonical ordering for mixing reader conditionals and regular clauses, e.g. by first sorting the pairs inside the reader conditional, sorting reader conditionals based on their sequences of pairs, and placing all reader conditionals before (or after) vanilla clauses.

I'm not sure it's worth trying to define alignment for reader conditionals, except maybe if there are multiple clauses inside of splicing conditionals.

File: how-to-ns/src/com/gfredericks/how_to_ns.clj (hunk 1 above)
Comment (vemv) [reply how-to-ns/src/com/gfredericks/how_to_ns.clj]: > x might be a reader conditional here, is that correct?

I scienced it and yes, correct. Will be fixing

**Comment by vemv** [comment]
Added support for various corner cases (reader conditionals placed in presumably unusual places), covered by tests and with per-commit granularity.

Turns out, supporting those could be as easy as upgrading `str` to `pr-str`, or similar minor tweaks.

Also fixed the corner case of `(:import (Thing))`, i.e. the `(Thing)` list having just one member. It was buggy in master even w/o reader conditionals.

Note that this branch has `npm` merged in - it will become cleaner after we are done with https://github.com/gfredericks/how-to-ns/pull/12

**Comment by gfredericks** [comment]
Seeing as

1. this problem is super complicated
2. I don't have immediate plans to make use of this feature
3. and I'm pretty busy
4. the prior two points mean I'm not very motivated to give it a lot of attention
5. but having the library not work at all when reader conditionals are around greatly reduces its utility

I'm going to merge this and add a disclaimer to the README saying that behavior with reader conditionals is subject to change.

Thanks for the work!

**Comment by gfredericks** [comment]
Released as `0.2.5`, thanks again.
